internal/provider/kubernetes/cache.go

```diff
@@ -0,0 +1,66 @@
+package kubernetes
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// providerCache maintains additional mappings related to Kubernetes provider
+// resources. The mappings are regularly updated from the reconcilers based
+// on the existence of the object in the Kubernetes datastore.
+type providerCache struct {
```

Comment:
Calling this a "cache" is misleading. It isn't really a cache, it's a reference-counting machine.